plugins/inputs/ethtool/ethtool_linux.go

```diff
@@ -143,41 +192,102 @@ func NewCommandEthtool() *CommandEthtool {
 }
 
 func (c *CommandEthtool) Init() error {
-	if c.ethtool != nil {
-		return nil
-	}
-
-	e, err := ethtoolLib.NewEthtool()
-	if err == nil {
-		c.ethtool = e
-	}
-
-	return err
+	return nil
 }
 
-func (c *CommandEthtool) DriverName(intf string) (string, error) {
-	return c.ethtool.DriverName(intf)
+func (c *CommandEthtool) DriverName(intf NamespacedInterface) (driver string, err error) {
+	return intf.Ethtool.DriverName(intf.Name)
 }
 
-func (c *CommandEthtool) Stats(intf string) (map[string]uint64, error) {
-	return c.ethtool.Stats(intf)
+func (c *CommandEthtool) Stats(intf NamespacedInterface) (stats map[string]uint64, err error) {
+	return intf.Ethtool.Stats(intf.Name)
 }
 
-func (c *CommandEthtool) Interfaces() ([]net.Interface, error) {
-	// Get the list of interfaces
+func (c *CommandEthtool) Interfaces() ([]NamespacedInterface, error) {
+	allInterfaces := make([]NamespacedInterface, 0)
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Get the list of interfaces in the init namespace
+	initialNamespace, err := netns.Get()
+	if err != nil {
+		return nil, err
+	}
+	e, err := ethtoolLib.NewEthtool()
+	if err != nil {
+		return nil, err
+	}
 	interfaces, err := net.Interfaces()
 	if err != nil {
 		return nil, err
 	}
+	for _, iface := range interfaces {
+		allInterfaces = append(
+			allInterfaces,
+			NamespacedInterface{
+				Interface:     iface,
+				Namespace:     initialNamespace,
+				NamespaceName: "",
+				Ethtool:       e,
+			},
+		)
+	}
+
+	namespaces, err := os.ReadDir("/var/run/netns")
+	if err != nil {
+		c.Log.Warnf("Could not find namespace directory: %s", err)
+		return allInterfaces, nil
+	}
+	for _, namespaceFile := range namespaces {
+		namespace, err := netns.GetFromName(namespaceFile.Name())
+		if err != nil {
+			c.Log.Warnf("Could not get namespace %s: %s", namespaceFile.Name(), err)
+			continue
+		}
+		if err := netns.Set(namespace); err != nil {
+			c.Log.Warnf("Could not switch to namespace %s: %s", namespaceFile.Name(), err)
+			continue
+		}
+		e, err := ethtoolLib.NewEthtool()
+		if err != nil {
+			c.Log.Warnf("Could not connect to ethtool in namespace %s: %s", namespaceFile.Name(), err)
+			continue
+		}
+
+		interfaces, err := net.Interfaces()
+		if err != nil {
+			c.Log.Warnf("Could not get interfaces in namespace %s: %s", namespaceFile.Name(), err)
+			continue
+		}
+		for _, iface := range interfaces {
+			allInterfaces = append(
+				allInterfaces,
+				NamespacedInterface{
+					Interface:     iface,
+					Namespace:     namespace,
+					NamespaceName: namespaceFile.Name(),
+					Ethtool:       e,
+				},
+			)
+		}
+	}
+
+	if err := netns.Set(initialNamespace); err != nil {
+		c.Log.Errorf("Could not return to initial namespace: %s", err)
+		return nil, err
+	}
```

Comment:
My concern with this is if an existing user has some additional network namespaces created, they will suddenly start getting errors from Telegraf around operation not permitted, right?

As an example, I created one namespace and ran again with no elevated permissions and error out, whereas before I was collecting metrics without issue:

```s
2022-09-28T14:39:04Z W! [inputs.ethtool] Could not switch to namespace vnet0: operation not permitted
2022-09-28T14:39:04Z E! [inputs.ethtool] Could not return to initial namespace: operation not permitted
2022-09-28T14:39:04Z E! [inputs.ethtool] Error in plugin: operation not permitted
```

Reply:
Hmm.. I did make some changes that I didn't put through my local testing to satisfy the CI linter. I wonder if that changed things. The intention is that unless at least one of the new namespace filters is configured, behavior should be unchanged from prior to this change.
Let me investigate...

Reply:
Found and fixed the bug. It should be working as intended now (extra warnings in the logs, but functions as before otherwise).